File: packages/flutter_duit/lib/src/duit_impl/action.dart
import 'package:flutter_duit/src/utils/index.dart';

final class HttpActionMetainfo {
  String method;

  HttpActionMetainfo({required this.method});

  static HttpActionMetainfo? fromJson(JSONObject? json) {
    if (json == null) return null;

    return HttpActionMetainfo(method: json["method"] ?? "GET");
  }
}

final class ActionDependency {
  String id;
  String target;

  ActionDependency({
    required this.target,
    required this.id,
  });

  factory ActionDependency.fromJSON(JSONObject json) {
    return ActionDependency(
      target: json["target"],
      id: json["id"],
    );
  }
}

final class ServerAction {
  List<ActionDependency> dependsOn;
  String event;
  HttpActionMetainfo? meta;

  ServerAction({
    required this.event,
    this.dependsOn = const [],
    this.meta,
  });

   static ServerAction? fromJSON(JSONObject? json) {
     if (json == null) return null;

    List<ActionDependency> deps = [];

    if (json["dependsOn"] != null) {
      json["dependsOn"].forEach((el) {
        final it = ActionDependency.fromJSON(el);
        deps.add(it);
      });
    }

    return ServerAction(
      event: json["event"] ?? "",
      dependsOn: deps,
      meta: HttpActionMetainfo.fromJson(json["meta"]),
    );
  }

  @override
  String toString() {
    return 'ServerAction{dependsOn: $dependsOn, event: $event, meta: $meta}';
  }
}

File: packages/duit_js/src/lib/action.ts
import { ID } from "../utils/id";

type HttpMethod = "POST" | "GET" | "PATCH" | "DELETE";

interface ActionDependency {
    id: ID;
    target: string;
}

interface HttpActionMetainfo {
    method: HttpMethod;
}

export class BaseAction {
    dependsOn: ActionDependency[];

    /**
     * Url (for http) of event name (for ws)
     */
    event: string;

    constructor(event: string, dependsOn: ActionDependency[] = []) {
        this.event = event;
        this.dependsOn = dependsOn;
    }
}


export class HttpAction extends BaseAction {
    meta: HttpActionMetainfo;

    constructor(event: string, meta: HttpActionMetainfo, dependsOn: ActionDependency[] = []) {
        super(event, dependsOn);
        this.meta = meta ?? {};
    }
}

export class WebSocketAction extends BaseAction {
    constructor(event: string, dependsOn: ActionDependency[] = []) {
        super(event, dependsOn);
    }
}
